src/verifier.cc

```diff
@@ -15,15 +15,73 @@
 #include "src/verifier.h"
 
 #include <cmath>
+#include <vector>
+
+#include "src/command.h"
 
 namespace amber {
 namespace {
 
-const float kEpsilon = 0.002f;
+const double kEpsilon = 0.002;
 
-bool IsFloatPixelEqualInt(float pixel, uint8_t expected) {
+bool IsFuzzyEqual(double real, double expected) {
   // TODO(jaebaek): Change kEpsilon to tolerance.
-  return std::fabs(pixel - static_cast<float>(expected) / 255.0f) < kEpsilon;
+  return std::fabs(real - expected) < kEpsilon;
+}
+
+template <typename T>
+Result CheckValue(ProbeSSBOCommand::Comparator comp,
+                  const uint8_t* memory,
+                  const std::vector<Value>& values) {
+  const T* ptr = reinterpret_cast<const T*>(memory);
+  for (auto v : values) {
+    const T val = v.IsInteger() ? static_cast<T>(v.AsUint64())
+                                : static_cast<T>(v.AsDouble());
+
+    switch (comp) {
+      case ProbeSSBOCommand::Comparator::kEqual:
+        if (*ptr != val)
+          return Result("Verifier failed: " + std::to_string(*ptr) +
+                        " != " + std::to_string(val));
+        break;
+      case ProbeSSBOCommand::Comparator::kNotEqual:
+        if (*ptr == val)
+          return Result("Verifier failed: " + std::to_string(*ptr) +
+                        " == " + std::to_string(val));
+        break;
+      case ProbeSSBOCommand::Comparator::kFuzzyEqual:
+        if (!IsFuzzyEqual(static_cast<double>(*ptr), val))
+          return Result("Verifier failed: " + std::to_string(*ptr) +
+                        " is not fuzzy-equal to " + std::to_string(val));
+        break;
+      case ProbeSSBOCommand::Comparator::kLess:
+        if (*ptr >= val)
+          return Result("Verifier failed: " + std::to_string(*ptr) +
+                        " >= " + std::to_string(val));
+        break;
+      case ProbeSSBOCommand::Comparator::kLessOrEqual:
+        if (*ptr > val)
+          return Result("Verifier failed: " + std::to_string(*ptr) + " > " +
+                        std::to_string(val));
+        break;
+      case ProbeSSBOCommand::Comparator::kGreater:
+        if (*ptr <= val)
+          return Result("Verifier failed: " + std::to_string(*ptr) +
+                        " <= " + std::to_string(val));
```

Comment:
For these error messages, I think we should use the comparator not the comparison used. So, this should be std::to_string(*ptr) + " > " + std::to_string(val); because that's the check in the script which failed.

Reply:
done